lib/ecmascript.ts

```diff
@@ -221,594 +213,299 @@ function getIntlDateTimeFormatEnUsForTimeZone(timeZoneIdentifier) {
-      offset,
-      z,
-      calendar
-    };
-  },
-  ParseTemporalInstantString: (isoString) => {
-    return ES.ParseISODateTime(isoString, { zoneRequired: true });
-  },
-  ParseTemporalZonedDateTimeString: (isoString) => {
-    return ES.ParseISODateTime(isoString, { zoneRequired: true });
-  },
-  ParseTemporalDateTimeString: (isoString) => {
-    return ES.ParseISODateTime(isoString, { zoneRequired: false });
-  },
-  ParseTemporalDateString: (isoString) => {
-    return ES.ParseISODateTime(isoString, { zoneRequired: false });
-  },
-  ParseTemporalTimeString: (isoString) => {
-    const match = PARSE.time.exec(isoString);
-    let hour, minute, second, millisecond, microsecond, nanosecond, calendar;
-    if (match) {
-      hour = ES.ToInteger(match[1]);
-      minute = ES.ToInteger(match[2] || match[5]);
-      second = ES.ToInteger(match[3] || match[6]);
-      if (second === 60) second = 59;
-      const fraction = (match[4] || match[7]) + '000000000';
-      millisecond = ES.ToInteger(fraction.slice(0, 3));
-      microsecond = ES.ToInteger(fraction.slice(3, 6));
-      nanosecond = ES.ToInteger(fraction.slice(6, 9));
-      calendar = match[15];
-    } else {
-      ({ hour, minute, second, millisecond, microsecond, nanosecond, calendar } = ES.ParseISODateTime(isoString, {
-        zoneRequired: false
-      }));
-    }
-    return { hour, minute, second, millisecond, microsecond, nanosecond, calendar };
-  },
-  ParseTemporalYearMonthString: (isoString) => {
-    const match = PARSE.yearmonth.exec(isoString);
-    let year, month, calendar, referenceISODay;
-    if (match) {
-      let yearString = match[1];
-      if (yearString[0] === '\u2212') yearString = `-${yearString.slice(1)}`;
-      year = ES.ToInteger(yearString);
-      month = ES.ToInteger(match[2]);
-      calendar = match[3];
-    } else {
-      ({ year, month, calendar, day: referenceISODay } = ES.ParseISODateTime(isoString, { zoneRequired: false }));
-    }
-    return { year, month, calendar, referenceISODay };
-  },
-  ParseTemporalMonthDayString: (isoString) => {
-    const match = PARSE.monthday.exec(isoString);
-    let month, day, calendar, referenceISOYear;
-    if (match) {
-      month = ES.ToInteger(match[1]);
-      day = ES.ToInteger(match[2]);
-    } else {
-      ({ month, day, calendar, year: referenceISOYear } = ES.ParseISODateTime(isoString, { zoneRequired: false }));
-    }
-    return { month, day, calendar, referenceISOYear };
-  },
-  ParseTemporalTimeZoneString: (stringIdent) => {
-    try {
-      let canonicalIdent = ES.GetCanonicalTimeZoneIdentifier(stringIdent);
-      if (canonicalIdent) {
-        canonicalIdent = canonicalIdent.toString();
-        if (ES.ParseOffsetString(canonicalIdent) !== null) return { offset: canonicalIdent };
-        return { ianaName: canonicalIdent };
-      }
-    } catch {
-      // fall through
+    !HasSlot(item, ISO_YEAR, ISO_MONTH, ISO_DAY)
+  );
+}
+export function IsTemporalDateTime(item) {
```

Comment:
I found it frustrating when reviewing that the old code and new code doesn't line up in the diff, even though almost all the code is identical. Could this be fixed by restoring the old indenting (indent all the new stuff two spaces) during the review of this PR?

Then right before merging you could add another commit to remove the extra indent. 

I'm not sure if this is enough to solve the diff problem, but seems worth a try to make it easier to review.

Reply:
Undocumented GitHub trick that someone once told me: you can append `?w=1` to the URL to print a diff that ignores whitespace and indentation changes. (It's still a large diff, but this cuts it down to about a third of the size, and the blocks match up better.)

Reply:
Wow that's a great trick!  It totally works!  @12wrigja, never mind my suggestion above: @ptomato's trick does the same thing without messing with commits. 

BTW, looks like there's a UI option for getting this behavior:
<img src=https://user-images.githubusercontent.com/277214/134785893-a29dce78-a9a8-4e12-a897-4b3322e251f7.png height=180>

Reply:
Ah yeah - I was playing around with that setting yesterday with the hope that this would make things a bit easier. There's at least a couple places where the diffs still doesn't line up much, but it does reduce the diffs down significantly.

That being said, I still plan to split out as many changes as possible and land those separately (e.g. most new type annotations, runtime changes etc) before sending this for review.

Reply:
> That being said, I still plan to split out as many changes as possible and land those separately (e.g. most new type annotations, runtime changes etc) before sending this for review.

Would it be easier to land this PR before putting type annotations on every exported function parameter?  I ask because it seems like TS seems to struggle with typing the ES object, perhaps because it has so many methods. On my fork, ES resolved to just `any`.  So I suspect it will be easier to find type errors if you first convert from `export const ES` => separate exports. But you're closer to this than me, so if you have a better plan feel free to do that instead!

I agree with fixing obvious type errors that show up simply as a result of doing the  `export const ES` => separate exports. Mostly thinking of the next-level-down typing that's not required to make this PR go green. (Except for the `any` ones that IMHO make sense to fix here.)

Reply:
As you mention, I plan to land this and then change the types. In many places, the type inference that tsc will now be able to do should help me greatly in deciding how to type things (though, especially for e.g. options bags if you know that there's a pre-defined type from e.g. the public API that is passed all the way through, then it would be awesome if you wanted to send a PR to use that once this lands).